File: packages/sdk/ios/core/BLEIntSDK/Classes/Protocol/ProtocolConstants.swift
//
//  ProtocolConstants
//
//  Copyright © 2022 2hire.io. All rights reserved.
//

import Foundation

internal let ProtocolVersion: UInt8 = 0x01

internal enum ProtocolFrame {
    static let SessionStart: [UInt8] = [0xC1, 0xA0, 0xC1, 0xA0]
    static let SessionEnd: [UInt8] = [0xE2, 0x1B, 0xE7, 0x7A]

    static let CommandStart: [UInt8] = [0xE2, 0x2A, 0xDA, 0xDA]
    static let CommandEnd: [UInt8] = [0xE2, 0x1B, 0xE7, 0x7A]
}

internal enum CommandIdentifier: UInt8 {
    case Ack = 0xF0
    case Nack = 0xF1
    case Error = 0xF2
}

internal enum ProtocolPacketType: UInt8 {
    case Request = 0xAA
    case Response = 0x55
}

internal enum ProtocolErrorCode: UInt8 {
    case AlreadyValidated = 0x01
    case InvalidAccess = 0xFF
    case InvalidTimestamp = 0xFE
    case InvalidTag = 0xFD
    case InvalidCounter = 0xFC
    case InvalidCmd = 0xFB
    case InvalidStartByte = 0xFA
    case UnknownVersion = 0xF9
    case InvalidCmdTag = 0xF8
    case InvalidLength = 0xF7
}

File: packages/sdk/android/core/src/main/java/io/twohire/bleintsdk/protocol/ProtocolConstants.kt
package io.twohire.bleintsdk.protocol

internal enum class ProtocolFrame(val rawValue: ByteArray) {
    SESSION_START(
        byteArrayOf(
            0xC1.toByte(), 0xA0.toByte(), 0xC1.toByte(), 0xA0.toByte()
        )
    ),
    SESSION_END(byteArrayOf(0xE2.toByte(), 0x1B.toByte(), 0xE7.toByte(), 0x7A.toByte())),
    COMMAND_START(byteArrayOf(0xE2.toByte(), 0x2A.toByte(), 0xDA.toByte(), 0xDA.toByte())),
    COMMAND_END(byteArrayOf(0xE2.toByte(), 0x1B.toByte(), 0xE7.toByte(), 0x7A.toByte()))
}

internal object ProtocolMessageType {
    const val REQUEST = 0xAA.toByte()
    const val RESPONSE = 0x55.toByte()
}

internal const val PROTOCOL_VERSION = 0x01.toByte()

internal enum class CommandIdentifier(val rawValue: Byte) {
    ACK(0xF0.toByte()),
    NACK(0xF1.toByte()),
    ERROR(0xF2.toByte());

    companion object {
        fun fromRaw(rawValue: Byte) = values().firstOrNull { it.rawValue == rawValue }
    }
}

internal enum class ProtocolPacketType(val rawValue: Byte) {
    REQUEST(0xAA.toByte()),
    RESPONSE(0x55.toByte());

    companion object {
        fun fromRaw(rawValue: Byte) = values().firstOrNull { it.rawValue == rawValue }
    }
}

internal enum class ProtocolErrorCode(val rawValue: Byte) {
    ALREADY_VALIDATED(0x01.toByte()),
    INVALID_ACCESS(0xFF.toByte()),
    INVALID_TIMESTAMP(0xFE.toByte()),
    INVALID_TAG(0xFD.toByte()),
    INVALID_COUNTER(0xFC.toByte()),
    INVALID_CMD(0xFB.toByte()),
    INVALID_START_BYTE(0xFA.toByte()),
    UNKNOWN_VERSION(0xF9.toByte()),
    INVALID_CMD_TAG(0xF8.toByte()),
    INVALID_LENGTH(0xF7.toByte());

    companion object {
        fun fromRaw(rawValue: Byte) = values().firstOrNull { it.rawValue == rawValue }
    }
}
